arch/arc/core/thread.c

```diff
@@ -64,14 +64,41 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 
 	char *stackEnd = pStackMem + stackSize;
 	struct init_stack_frame *pInitCtx;
-
+#if CONFIG_USERSPACE
+	/* for kernel thread, the privilege stack is merged into thread stack */
+	if (!(options & K_USER)) {
+	/* if MPU_STACK_GUARD is enabled, reserve the the stack area
+	 * |---------------------|----------------|
+	 * |                     |(MPU STACK AREA)|
+	 * |			 |----------------|
+	 * |  user stack         | kernel thread  |
+	 * |---------------------| stack          |
+	 * |  privilege stack    |                |
+	 * ----------------------------------------
```

Comment:
Not sure I understand this diagram. Can you make it linear?
Under what circumstances is memory arranged like this, rather than the diagram in arc_core_mpu.c?

Reply:
According to current macros in arch.h,  for a kernel thread,  a privilege stack will be allocated, like this 

thread_stack[stacksize + CONFIG_PRIVILEGED_STACK_SIZE].

So I adjust the thread sp and merge the privilege stack with user defined stack,  also reserve an area for possible mpu stack area.

So for a kernel thread,   a size of stacksize + CONFIG_PRIVILEGED_STACK_SIZE - MPU_STACK_GUARD_SIZE (> stacksize) stack can be used.

Reply:
As noted in my other comment I'd like to see this implementation align with what ARM is doing.